aten/src/ATen/native/DilatedConvolution.cpp

```diff
@@ -0,0 +1,1090 @@
+      stride_size,                      \
+      pad_size,                         \
+      dilation_size)
+
+#define CALL_BACKWARD_OUT(DIM)           \
+  conv_dilated##DIM##d_backward_out_cpu( \
+      grad_input,                        \
+      grad_weight,                       \
+      grad_bias,                         \
+      grad_output,                       \
+      input,                             \
+      weight,                            \
+      kernel_size,                       \
+      stride_size,                       \
+      pad_size,                          \
+      dilation_size,                     \
+      columns,                           \
+      ones)
+
+#define INSERT_BATCH_DIMENSION(A, SIZE)        \
+  if (A.numel() > 0) {                         \
+    auto new_sizes = A.sizes().vec();          \
+    new_sizes.insert(new_sizes.begin(), SIZE); \
+    A.resize_(new_sizes);                      \
+  }
+
+#define DROP_BATCH_DIMENSION(A)    \
+  if (A.numel() > 0) {             \
+    A.resize_(A.sizes().slice(1)); \
+  }
+
+namespace at {
+namespace native {
+
+namespace {
+
+inline bool all_positive(IntArrayRef& arr) {
+  return std::all_of(
+      arr.begin(), arr.end(), [](int64_t item) { return item > 0; });
+}
+
+#define OUTPUTSIZE(INDEX, DIM)                                       \
+  (input.size((INDEX) + input.dim() - (DIM)) + 2 * pad_size[INDEX] - \
+   (dilation_size[INDEX] * (kernel_size[INDEX] - 1) + 1)) /          \
+          stride_size[INDEX] +                                       \
+      1
+
+void conv_dilated_shape_check(
+    int64_t dim,
+    const Tensor& input,
+    const Tensor& weight,
+    const Tensor& bias,
+    const Tensor& grad_output,
+    const Tensor& grad_weight,
+    const Tensor& grad_bias,
+    IntArrayRef kernel_size,
+    IntArrayRef stride_size,
+    IntArrayRef pad_size,
+    IntArrayRef dilation_size,
+    const Tensor& columns,
+    const Tensor& ones) {
+  TORCH_CHECK(
+      kernel_size.size() == dim,
+      "kernel sizes length should be ",
+      dim,
+      ", but got ",
+      kernel_size.size());
+  TORCH_CHECK(
+      stride_size.size() == dim,
+      "strides length should be ",
+      dim,
+      ", but got ",
+      stride_size.size());
+  TORCH_CHECK(
+      dilation_size.size() == dim,
+      "dilations length should be ",
+      dim,
+      ", but got ",
+      dilation_size.size());
+  TORCH_CHECK(
+      pad_size.size() == dim,
+      "pads length should be ",
+      dim,
+      ", but got ",
+      pad_size.size());
+
+  TORCH_CHECK(
+      all_positive(kernel_size),
+      "kernel size should be greater than zero, but got ",
+      kernel_size);
+  TORCH_CHECK(
+      all_positive(stride_size),
+      "stride should be greater than zero, but got ",
+      stride_size);
+  TORCH_CHECK(
+      all_positive(dilation_size),
+      "dilation should be greater than zero, but got ",
+      dilation_size);
+
+  if (weight.numel() > 0) {
```

Comment:
You should use undefined tensor as the sentinel for "NULL" tensor pointer, not a zero-size tensor. (A zero-size tensor may be legitimately get passed in by the user, and your error checking code will get bamboozled in this case.)

Reply:
I'm not going to mention where other cases of this happen but there are quite a few of them.

Reply:
Right, now using `.defined()` instead of `.numel()>0`.